File: packages/pluto/src/clients/aws/bucket.s3.ts
import * as fs from "fs";
import assert from "assert";
import { promisify } from "util";
import { Readable, finished } from "stream";
import { S3Client, PutObjectCommand, GetObjectCommand } from "@aws-sdk/client-s3";
import { genResourceId } from "@plutolang/base/utils";
import { IBucketClient, Bucket, BucketOptions } from "../../bucket";
import { genAwsResourceName } from "./utils";

const promisifiedFinish = promisify(finished);

/**
 * Implementation of Bucket using S3.
 */
export class S3Bucket implements IBucketClient {
  private readonly id: string;
  private readonly bucketName: string;

  private client: S3Client;

  constructor(name: string, opts?: BucketOptions) {
    this.id = genResourceId(Bucket.fqn, name);
    this.bucketName = genAwsResourceName(this.id);

    this.client = new S3Client();
    opts;
  }

  public async put(fileKey: string, filePath: string): Promise<void> {
    const fileContent = fs.readFileSync(filePath);

    const uploadParams = {
      Bucket: this.bucketName,
      Key: fileKey,
      Body: fileContent,
    };

    await this.client.send(new PutObjectCommand(uploadParams));
  }

  public async get(fileKey: string, filePath: string): Promise<void> {
    const downloadParams = {
      Bucket: this.bucketName,
      Key: fileKey,
    };

    const data = await this.client.send(new GetObjectCommand(downloadParams));
    assert(data.Body, "data.Body is undefined");
    const fileStream = fs.createWriteStream(filePath);
    const stream = (data.Body as Readable).pipe(fileStream);
    await promisifiedFinish(stream);
  }
}

File: packages/pluto-py/pluto_client/clients/aws/bucket_s3.py
import boto3
from typing import Optional
from pluto_base import utils
from botocore.exceptions import NoCredentialsError, ClientError
from ...bucket import Bucket, BucketOptions, IBucketClient
from .utils import gen_aws_resource_name


class S3Bucket(IBucketClient):
    def __init__(self, name: str, opts: Optional[BucketOptions] = None):
        self.__id = utils.gen_resource_id(Bucket.fqn, name)
        self.__bucket_name = gen_aws_resource_name(self.__id)
        self.__client = boto3.client("s3")

    def put(self, file_key: str, file_path: str):
        try:
            self.__client.upload_file(file_path, self.__bucket_name, file_key)
        except FileNotFoundError:
            raise FileNotFoundError("The file was not found.")
        except NoCredentialsError:
            raise

    def get(self, file_key: str, file_path: str):
        try:
            self.__client.download_file(self.__bucket_name, file_key, file_path)
        except ClientError as e:
            if e.response["Error"]["Code"] == "404":
                raise FileNotFoundError("The object does not exist.")
            else:
                raise
